storages/response_picker.go

```diff
@@ -0,0 +1,148 @@
+package storages
+
+import (
+	"net/http"
+)
+
+// BasePicker contains common methods of pickers
+type BasePicker struct {
+	responsesChan <-chan BackendResponse
+	success       BackendResponse
+	failure       BackendResponse
+	errors        []BackendResponse
+	sent          bool
+}
+
+func (orp *BasePicker) collectSuccessResponse(bresp BackendResponse) {
+	if orp.hasSuccessfulResponse() {
+		bresp.DiscardBody()
+	} else {
+		orp.success = bresp
+	}
+}
+
+func (orp *BasePicker) collectFailureResponse(bresp BackendResponse) {
+	if orp.hasFailureResponse() {
+		bresp.DiscardBody()
+	} else {
+		orp.failure = bresp
+	}
+	orp.errors = append(orp.errors, bresp)
+}
+
+func (orp *BasePicker) hasSuccessfulResponse() bool {
+	return orp.success != BackendResponse{}
```

Comment:
Maybe we can keep `BackendResponse{}` in some kind of a constant called `emptyResponse`?